llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -211,43 +213,118 @@ EntryPointGroupVec groupEntryPointsByScope(const Module &M,
 
       Attribute Id = F.getFnAttribute(ATTR_SYCL_MODULE_ID);
       StringRef Val = Id.getValueAsString();
-      EntryPointMap[Val].push_back(&F);
+      EntryPointMap[Val].insert(&F);
       break;
     }
 
     case Scope_Global:
       // the map key is not significant here
-      EntryPointMap[GLOBAL_SCOPE_NAME].push_back(&F);
+      EntryPointMap[GLOBAL_SCOPE_NAME].insert(&F);
       break;
     }
   }
 
   if (!EntryPointMap.empty()) {
     EntryPointGroups.reserve(EntryPointMap.size());
-    for (auto &EPG : EntryPointMap)
-      EntryPointGroups.push_back({EPG.first, std::move(EPG.second)});
+    for (auto &EPG : EntryPointMap) {
+      EntryPointGroups.emplace_back(EntryPointGroup{
+          EPG.first, std::move(EPG.second), MD.getEntryPointGroup().Props});
+      EntryPointGroup &G = EntryPointGroups.back();
+      G.Props.Scope = EntryScope;
+    }
+  } else {
+    // No entry points met, record this.
+    EntryPointGroups.emplace_back(
+        EntryPointGroup{GLOBAL_SCOPE_NAME, EntryPointSet{}});
+  }
+  return EntryPointGroups;
+}
+
+template <class EntryPoinGroupFunc>
```

Comment:
Is this supposed to be EntryPointGroupFunc. It might be my lack of c++ knowhow, but why do we need a template here?

Reply:
EntryPointGroupFunc is the type of 'F' on the line 247. Template here is needed because lambda function is passed to here (e.g. L674 in this file). IMO, this place is correct.

I have other comment, but it is non-blocking. I personally don't like re-using the name 'F' for the function parameter at L247 and 'F' at L253, but that is not critical as the code is correct as is.

Reply:
Thanks @v-klochkov

Reply:
Nit pick. Can be addressed later

Reply:
Thanks folks for the review and merge while I was away. I'll examine all the "Can be addressed later" comments and create a separate patch.